memcpy: fix bug in arch_memcpy_s

mrajwa opened the pull request:
The current implementation of arch_memcpy_s() doesn't
allow to copy from adjecent memory block. This patch fixes
this problem.

tlauda commented on src/audio/kpb.c:
```diff
@@ -551,8 +551,7 @@ static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source,
 			 */
-			assert(!memcpy_s(buff->w_ptr, space_avail, read_ptr,
-					 space_avail));
+			memcpy(buff->w_ptr, read_ptr, space_avail);
```
@mrajwa Could you give any specific example? memcpy_s doesn't support wrapping automatically, so you still shouldn't fail at any time.

mrajwa replied:
@tlauda as you know history buffer uses large continuous blocks from various memory pools. Let's take HPSRAM as an example here. We buffer data internally.... at some point we have a situation like:

write pointer (history buffer) -> 0xbe01f480, 
read pointer (source component buffer ) -> 0xbe01f980, 
size to copy -> 1280 bytes.

now quick reminder of memcpy_s check:
`if ((dest + dest_size >= src && dest + dest_size <= src + src_size)`

Calculating if condition:

dest + dest_size = 0xbe01f480 + 1280 = 0xbe01f980
src + src_size = 0xbe01f980 + 1280  = 0xbe01fe80

so finally substituting values into if condition:

`if (TRUE && TRUE) `

This corner case where source component buffer happen to be right next to the end of our history buffer but such is life :)

mrajwa replied:
Although I think instead of removing these asserts and "safe" memcpy I would just go and allocate a byte more memory to avid such narrow cases.

tlauda replied:
@mrajwa It looks like a workaround and not a fix. From your example it's clearly the wrong condition for buffer overlapping. I would go with something like this:
`if ((dest > src && dest < src + src_size) || (src > dest && src < dest + dest_size))`

mrajwa replied:
@tlauda, you just dropped the equal conditions but they are actually fine. After all source and destination should not be the same...assuming any reasonable copy right?

tlauda replied:
@mrajwa I've proposed a little bit different condition. Your destination is ending at 0xbe01f980 and your source is starting at the same address. They are not overlapping, so where do you see the problem?

mrajwa replied:
@tlauda it is the question if we want actions like this:
`memcpy_s(0xbe01f980, 0x10, 0xbe01f980, 0x10)`

to raise an error or not. With your proposal it will be treated as permitted copy. Although it will not cause any issue it is a bit weird.

jajanusz replied:
> @tlauda it is the question if we want actions like this:
> `memcpy_s(0xbe01f980, 0x10, 0xbe01f980, 0x10)`
> 
> to raise an error or not. With your proposal it will be treated as permitted copy. Although it will not cause any issue it is a bit weird.

Overlapping source and destination have to raise error. It's like that in standard. Example that you given are overlapping buffers so yea, it's error.

jajanusz replied:
@tlauda @lgirdwood BTW shouldn't we have `restrict` in memcpys, to have it like in standard and maybe to get a performance (at least xtensa likes it)?

keyonjie replied:
@mrajwa @lgirdwood this is the bug I mentioned to you in the email, let me submit a patch for it.

keyonjie replied:
let's keep this assert usage here.

tlauda commented:
@jajanusz Yes, it wouldn't hurt and probably help with gcc optimization. I'm not sure if it's going to do anything for xcc, since we are already using vec_memcpy.
@mrajwa Maybe we should add additional check for src == dst. You think such final condition will handle all corner cases?

mrajwa commented:
> Maybe we should add additional check for src == dst

@tlauda and treat such condition as permitted?

tlauda commented:
@mrajwa Not permitted.

jajanusz commented:
@tlauda @mrajwa IDK if your problem is KPB or memcpy_s implementation. But memcpy_s is broken, if there fails for adjacent buffers if dest is before src ...

tlauda commented:
@jajanusz Exactly. Condition for buffer overlap needs to be fixed in memcpy_s.

mrajwa commented:
@jajanusz @tlauda yes obviously `arch_memcpy_s()` implementation has a bug

jajanusz commented on src/arch/host/include/arch/string.h:
```diff
@@ -32,8 +32,11 @@ static inline int arch_memcpy_s(void *dest, size_t dest_size,
 
-	if ((dest + dest_size >= src && dest + dest_size <= src + src_size) ||
-		(src + src_size >= dest && src + src_size <= dest + dest_size))
+	if (dest == src)
```
Need to do the same in `test/cmocka/src/universal_mocks.c`

jajanusz commented on src/arch/host/include/arch/string.h:
```diff
@@ -32,8 +32,11 @@ static inline int arch_memcpy_s(void *dest, size_t dest_size,
+	if (dest == src)
+		return -EINVAL;
+
+	if ((dest + dest_size > src && dest + dest_size < src + src_size) ||
```
It's still invalid f.e. it should fail for that case:
```c
char a[8] = {0};
assert_int_not_equal(memcpy_s(a, 2, a + 1, 1), 0);
```
Also there are unnecessary additions in current implementation.
I think it's the correct solution:
```c
if ((dest >= src && dest < (src + src_size)) ||
		(src >= dest && src < (dest + dest_size)))
```

jajanusz replied:
Also with code above check for dst==src is not needed, however with 0-sized arrays it will pass this check, but it shouldn't do anything with that size anyway

mrajwa replied:
> however with 0-sized arrays it will pass this check

Because this is the bare minimum we should know and check ourselves.

> It's still invalid f.e. it should fail for that case:
> 
> ```c
> char a[8] = {0};
> assert_int_not_equal(memcpy_s(a, 2, a + 1, 1), 0);
> ```

> Please use @jajanusz's solution.

I don't think it should fail in this case. You say "array `a` has two bytes, please copy one byte from address `a+1` to the address `a`" Sounds perfectly legal to me, don't you think?
What your memcpy_s do is in essence a[0] = a[1] -> Why would we forbid such copy?

jajanusz replied:
> Sounds perfectly legal to me, don't you think?

We may think whatever we want, but standard says clearly that buffers in memcpy **cannot** overlap. These arguments are invalid. So yea it's an error.

mrajwa replied:
@jajanusz but they don't overlap in your example, do they? Also what exactly do you read? I guess you read memcpy specification but here we discuss memcpy_s, a wrapper function.

jajanusz replied:
> but they don't overlap in your example, do they?

last byte of dst is first byte of src, they overlap on tha 1 byte





> Also what exactly do you read? I guess you read memcpy specification but here we discuss memcpy_s, a wrapper function.

It's invalid for both. For `memcpy` it results with *undefined behavior*, for `memcpy_s` it's an error.

tlauda commented:
Please use @jajanusz's solution.

jajanusz commented:
@mrajwa Please update, memcpy bug sounds kinda critical

mrajwa commented:
@jajanusz @lgirdwood, updated. However I must say I really don't like this memcpy_s(). First with current implementation we fixed the allocation of adjacent memory blocks but you can use it copy bytes inside one mmeory block. Whether a block returned by allocator or simple array you can't do equivalent of `a[0] = a[1]` with this memcpy_s. It treats such **internal** copies as overlapping. Second it is not very **secure** to me. We have to provide two sizes now instead of one as in plain memcpy case. This alone opens a room for mistake. I think far better solution would be to implement more OO apprach with memory blocks datatype where tou keep all the informations about the block (sizes, start/end addresses etc).

jajanusz commented:
@mrajwa 
1) Subset is still overlapping. If overlapping itself is good or not, doesn't matter here as some1 who calls function expects it to behave as in standard...
2) You provide 2 sizes to prevent situations where src will not fit in dest, it may be troublesome, but at least it's extra check.
OO-aproach may be good in cases where you want to treat buffers as arrays, but it's not always the case. So you would probably need to do something more than just struct with size and pointer. Messing with that in C may not be nice and kernel developers usually perfer usual approach anyway ...
